Hmis loader performance

amichal opened the pull request:
- [x] Switch the initial load! from `activerecord-import` to PG `COPY DATA`  while still  preprocessing lines/headers in ruby to guess encoding and inject meta data  (~5x improvement on my benchmark)
- [x] Add detailed timing information to the debug logs
- [x] Remove preprocessed line counts, since the CSV can contain embedded newlines they don't tell us much
- [x] Cut down on allocations, precompute what we can and avoid newing up AR models where we can on import! (~3x improvement on my benchmark)
- [x] Tamper with `fix_time_format` used by importer to return Time/DateTime objects so import! does not screw up the default timezone

amichal committed:
Avoid reading the files multiple times, switch to PG COPY DATA to load! hmis_twenty_twenty/hud_sample in 16s instead of 1:40s on my machine

amichal committed:
Logger adjustments

amichal committed:
We were missing the first line of data

amichal committed:
This slows down queries, add a way to disable it

amichal committed:
WIP: Reduce CPU and allocation overhead. A roughly 3x perf improvement (with validations disabled for now)

amichal committed:
Fix? error paths

amichal committed:
Avoid the boundry case at the edge of the year, we dont which choice is made in the final hours of the year

amichal committed:
puts no longer needed

amichal committed:
Put row validations back to prove tests pass

amichal committed:
Start the switch to working with hashes in preprocess

amichal committed:
Merge branch 'pre-release' into hmis-loader-performance

amichal committed:
Start switching to returning a ActiveSupport::TimeWithZone so the faster importer has the timezone info

amichal committed:
Use TimeWithZone or DateTime always from fix_time_format

amichal committed:
Remove the final (for now) batch of FIXMEs

eanders committed:
Merge branch 'pre-release' into hmis-loader-performance

amichal committed:
Refactor load_source_file_pg so that with minimal changes we can switch back to INSERTs for verification. Fix record counting logic

amichal commented:
Testing  this weekend with more real world data found a few things I'm fixing.

amichal committed:
Merge branch 'pre-release' into hmis-loader-performance

eanders committed:
Merge branch 'pre-release' into hmis-loader-performance

eanders committed:
Merge branch 'hmis-loader-performance' of github.com:greenriver/hmis-warehouse into hmis-loader-performance

eanders committed:
Merge branch 'pre-release' into hmis-loader-performance

amichal committed:
Tidy up logger/notifier

eanders commented on drivers/hmis_csv_twenty_twenty/app/models/hmis_csv_twenty_twenty/importer/import_concern.rb:
```diff
@@ -205,20 +213,22 @@ def self.fix_date_format(string)
+    # the intended timezone otherwise
     def self.fix_time_format(string)
       return unless string
 
       string = string.gsub('/', '-')
       # Ruby handles yyyy-m-d just fine, so we'll allow that even though it doesn't match the spec
       return string if /\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}/.match?(string)
-      return string.to_time.strftime('%Y-%m-%d %H:%M:%S') if /\d{4}-\d{1,2}-\d{1,2} \d{1,2}:\d{1,2}:?\d{0,2}?/.match?(string)
+      return Time.zone.parse(string) if /\d{4}-\d{1,2}-\d{1,2} \d{1,2}:\d{1,2}:?\d{0,2}?/.match?(string)
 
       # Sometimes times come in mm-dd-yyyy hh:mm
       if /\d{1,2}-\d{1,2}-\d{4} \d{1,2}:\d{1,2}:?\d{0,2}?/.match?(string)
         date, time = string.split(' ')
         month, day, year = date.split('-')
-
-        return "#{year}-#{month.rjust(2, '0')}-#{day.rjust(2, '0')} #{time}".to_time.strftime('%Y-%m-%d %H:%M:%S')
+        return Time.zone.parse("#{year}-#{month.rjust(2, '0')}-#{day.rjust(2, '0')} #{time}")
```
I think this isn't picking the right timezone.  On the left, previous behavior, on the right, current behavior (loaded on top, imported on bottom)
![Screen Shot 2021-02-03 at 9 22 31 AM](https://user-images.githubusercontent.com/1346876/106760256-97a36380-6601-11eb-8b36-a2d11d33c6cf.png)

eanders committed:
Adjustment to the importer log format for slack

amichal commented:
Ugh... Maybe a testing, `travel_to` interaction with AR::Import? Not sure
why else it passed tests locally which look for this.

On Wed, Feb 3, 2021 at 9:24 AM Elliot <notifications@github.com> wrote:

> *@eanders* commented on this pull request.
> ------------------------------
>
> In
> drivers/hmis_csv_twenty_twenty/app/models/hmis_csv_twenty_twenty/importer/import_concern.rb
> <https://github.com/greenriver/hmis-warehouse/pull/1308#discussion_r569454710>
> :
>
> > +    # the intended timezone otherwise
>      def self.fix_time_format(string)
>        return unless string
>
>        string = string.gsub('/', '-')
>        # Ruby handles yyyy-m-d just fine, so we'll allow that even though it doesn't match the spec
>        return string if /\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}/.match?(string)
> -      return string.to_time.strftime('%Y-%m-%d %H:%M:%S') if /\d{4}-\d{1,2}-\d{1,2} \d{1,2}:\d{1,2}:?\d{0,2}?/.match?(string)
> +      return Time.zone.parse(string) if /\d{4}-\d{1,2}-\d{1,2} \d{1,2}:\d{1,2}:?\d{0,2}?/.match?(string)
>
>        # Sometimes times come in mm-dd-yyyy hh:mm
>        if /\d{1,2}-\d{1,2}-\d{4} \d{1,2}:\d{1,2}:?\d{0,2}?/.match?(string)
>          date, time = string.split(' ')
>          month, day, year = date.split('-')
> -
> -        return "#{year}-#{month.rjust(2, '0')}-#{day.rjust(2, '0')} #{time}".to_time.strftime('%Y-%m-%d %H:%M:%S')
> +        return Time.zone.parse("#{year}-#{month.rjust(2, '0')}-#{day.rjust(2, '0')} #{time}")
>
> I think this isn't picking the right timezone. On the left, previous
> behavior, on the right, current behavior (loaded on top, imported on bottom)
> [image: Screen Shot 2021-02-03 at 9 22 31 AM]
> <https://user-images.githubusercontent.com/1346876/106760256-97a36380-6601-11eb-8b36-a2d11d33c6cf.png>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/greenriver/hmis-warehouse/pull/1308#pullrequestreview-582417148>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEJOZVXVK7M527MOVZTLZDS5FMCRANCNFSM4WPEY5KQ>
> .
>

amichal commented:
I'll look into it.

On Wed, Feb 3, 2021 at 9:27 AM Aaron Michal <amichal@greenriver.org> wrote:

> Ugh... Maybe a testing, `travel_to` interaction with AR::Import? Not sure
> why else it passed tests locally which look for this.
>
> On Wed, Feb 3, 2021 at 9:24 AM Elliot <notifications@github.com> wrote:
>
>> *@eanders* commented on this pull request.
>> ------------------------------
>>
>> In
>> drivers/hmis_csv_twenty_twenty/app/models/hmis_csv_twenty_twenty/importer/import_concern.rb
>> <https://github.com/greenriver/hmis-warehouse/pull/1308#discussion_r569454710>
>> :
>>
>> > +    # the intended timezone otherwise
>>      def self.fix_time_format(string)
>>        return unless string
>>
>>        string = string.gsub('/', '-')
>>        # Ruby handles yyyy-m-d just fine, so we'll allow that even though it doesn't match the spec
>>        return string if /\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}/.match?(string)
>> -      return string.to_time.strftime('%Y-%m-%d %H:%M:%S') if /\d{4}-\d{1,2}-\d{1,2} \d{1,2}:\d{1,2}:?\d{0,2}?/.match?(string)
>> +      return Time.zone.parse(string) if /\d{4}-\d{1,2}-\d{1,2} \d{1,2}:\d{1,2}:?\d{0,2}?/.match?(string)
>>
>>        # Sometimes times come in mm-dd-yyyy hh:mm
>>        if /\d{1,2}-\d{1,2}-\d{4} \d{1,2}:\d{1,2}:?\d{0,2}?/.match?(string)
>>          date, time = string.split(' ')
>>          month, day, year = date.split('-')
>> -
>> -        return "#{year}-#{month.rjust(2, '0')}-#{day.rjust(2, '0')} #{time}".to_time.strftime('%Y-%m-%d %H:%M:%S')
>> +        return Time.zone.parse("#{year}-#{month.rjust(2, '0')}-#{day.rjust(2, '0')} #{time}")
>>
>> I think this isn't picking the right timezone. On the left, previous
>> behavior, on the right, current behavior (loaded on top, imported on bottom)
>> [image: Screen Shot 2021-02-03 at 9 22 31 AM]
>> <https://user-images.githubusercontent.com/1346876/106760256-97a36380-6601-11eb-8b36-a2d11d33c6cf.png>
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/greenriver/hmis-warehouse/pull/1308#pullrequestreview-582417148>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAEJOZVXVK7M527MOVZTLZDS5FMCRANCNFSM4WPEY5KQ>
>> .
>>
>

amichal committed:
Add some docs on why we and how we need to fixup HUD HMIS timestamps

amichal committed:
Trying strptime as maybe a fast and safe alternative

amichal committed:
A bunch of strings that might be date but arent

eanders committed:
Merge branch 'pre-release' into hmis-loader-performance

eanders committed:
Merge branch 'pre-release' into hmis-loader-performance

eanders committed:
New schema file

amichal committed:
Merge branch 'hmis-loader-performance' of github.com:greenriver/hmis-warehouse into hmis-loader-performance

eanders committed:
Merge branch 'pre-release' into hmis-loader-performance

eanders commented on drivers/hmis_csv_twenty_twenty/app/models/hmis_csv_twenty_twenty/importer/import_concern.rb:
```diff
@@ -167,102 +167,83 @@ def self.accepted_time_pattern
+    # current year so we choose an interpretation of a 2 digit year
+    # in that range on import if we have to.
+
+    RE_YYYYMMDD = /(?<y>\d{4})-(?<m>\d{1,2})-(?<d>\d{1,2})/.freeze
+
+    DATE_FORMATS = [
+      ['%Y-%m-%d', RE_YYYYMMDD],
+      ['%m-%d-%Y'],
+      ['%d-%b-%Y'],
+    ].freeze
+
+    TIME_FORMATS = ([
+      ['%Y-%m-%d %H:%M:%S', RE_YYYYMMDD],
+      ['%m-%d-%Y %H:%M:%S'],
+      ['%d-%b-%Y %H:%M:%S'],
+      ['%Y-%m-%d %H:%M', RE_YYYYMMDD],
+      ['%m-%d-%Y %H:%M'],
+      ['%d-%b-%Y %H:%M'],
+    ] + DATE_FORMATS).freeze # order matters, we need to try more logical and longer patterns first
```
This is a great idea.

amichal committed:
Avoid stomping constants, fix tests

amichal committed:
Merge branch 'hmis-loader-performance' of github.com:greenriver/hmis-warehouse into hmis-loader-performance

eanders commented on drivers/hmis_csv_twenty_twenty/app/models/hmis_csv_twenty_twenty/loader/loader.rb:
```diff
@@ -300,19 +327,23 @@ def export_file_valid?
+        status: status,
+      )
+      status = "#{status} error:#{err}" if err
+      log("Completed in #{elapsed_time(elapsed)} #{hash_as_log_str log_ids}. status:#{status}\n#{summary_as_log_str loader_log.summary}")
```
Let's make this `Completed loading in ....`

eanders commented on drivers/hmis_csv_twenty_twenty/app/models/hmis_csv_twenty_twenty/importer/importer.rb:
```diff
@@ -592,26 +678,39 @@ def self.soft_deletable_sources
+      importer_log.upload_id = @upload.id if @upload.present?
+      importer_log.save
+      elapsed = Time.current - @started_at
+      log("Completed in #{elapsed_time(elapsed)} #{hash_as_log_str log_ids}.\n#{summary_as_log_str importer_log.summary}")
```
Let's make this `Completed importing in ....`

amichal committed:
Small text change

eanders committed:
Merge branch 'pre-release' into hmis-loader-performance

amichal committed:
Validations should still be visible even for soft-deleted incoming records

amichal committed:
Empty strings need to mean NULL for most columns

amichal committed:
Use timezone aware date/time formatting always

amichal committed:
Merge branch 'hmis-loader-performance' of github.com:greenriver/hmis-warehouse into hmis-loader-performance

amichal committed:
Allow 1..2 years in the future, like the old code did

amichal committed:
Make sure we are testing in the app timezone and that we cover the edge cases in our century inference

eanders committed:
Merge branch 'pre-release' into hmis-loader-performance

eanders committed:
Smaller pagination

eanders committed:
Merge branch 'pre-release' into hmis-loader-performance
